core/src/main/java/org/infinispan/protostream/impl/TagWriterImpl.java

```diff
@@ -1191,13 +1382,82 @@ private int writeVarInt64Direct(int i, long value) throws IOException {
          return i;
       }
 
-      private static int varIntBytes(long value) {
-         int i = 1;
-         while ((value & 0xFFFFFFFFFFFFFF80L) != 0) {
-            ++i;
-            value >>>= 7;
+      @Override
+      Encoder subEncoder(int number) throws IOException {
+         // We assume worse case for tag and 1 byte for length
+         out.ensureCapacity(10 + 1);
+         int pos = out.getPosition();
+         pos = writeVarInt32Direct(pos, WireType.makeTag(number, WireType.WIRETYPE_LENGTH_DELIMITED));
+         out.setPosition(pos + 1);
+         // We just skip a byte, assuming the nested message is less than 128. If it isn't we will move it later
+         return new ChildOutputStreamRandomAccessEncoder(out);
+      }
+
+      @Override
+      public void close() throws IOException {
+         // This method should only be invoked for outer encoders only!
+         super.close();
+         out.reset();
+      }
+   }
+
+   static int varIntBytes(long value) {
+      int i = 1;
+      while ((value & 0xFFFFFFFFFFFFFF80L) != 0) {
+         ++i;
+         value >>>= 7;
+      }
+      return i;
+   }
+
+   static class ChildOutputStreamRandomAccessEncoder extends OutputStreamRandomAccessEncoder {
+      private int[] positions;
+      private int head;
+
+      private ChildOutputStreamRandomAccessEncoder(RandomAccessOutputStream raos) {
+         super(raos);
+      }
+
+      @Override
+      boolean canReuseWriter() {
+         return true;
+      }
+
+      @Override
+      Encoder subEncoder(int number) throws IOException {
+         out.ensureCapacity(6);
+         int pos = out.getPosition();
+         pos = writeVarInt32Direct(pos, WireType.makeTag(number, WireType.WIRETYPE_LENGTH_DELIMITED));
+         out.setPosition(pos + 1);
+         if (positions == null) {
+            positions = new int[10];
+         } else if (head == positions.length) {
+            int newSize = head + 10;
+            if (newSize > 50) {
```

Comment:
The max nested is configurable in `org.infinispan.protostream.config.Configuration#maxNestedMessageDepth()`
This check can be removed IMO.

Reply:
This is something I didn't look more closely into, but we don't have access to the config at this spot. The reason the check is needed is because this code uses the new `canReuseWriter` method and thus we only create a two TagWriterImpl instances per marshalling call (one for the outer and one for the child that handles all the nested messages). I had debated about passing down the max depth to the encoder, which I should probably do I guess.